lib/solaar/ui/pair_window.py

```diff
@@ -27,144 +28,104 @@
+    if receiver.receiver_kind == "bolt":
+        text = _("Bolt receivers are only compatible with Bolt devices.")
+        text += "\n\n"
+        text += _("Press a pairing button or key until the pairing light flashes quickly.")
+    else:
+        if receiver.receiver_kind == "unifying":
+            text = _("Unifying receivers are only compatible with Unifying devices.")
+        else:
+            text = _("Other receivers are only compatible with a few devices.")
+        text += "\n\n"
+        text += _("Turn on the device you want to pair.")
+        text += _("The device must not be paired with a nearby powered-on receiver.")
+        text += "\n"
+        text += _("If the device is already turned on, turn it off and on again.")
+    if receiver.remaining_pairings() and receiver.remaining_pairings() >= 0:
+        text += (
+            ngettext(
+                "\n\nThis receiver has %d pairing remaining.",
+                "\n\nThis receiver has %d pairings remaining.",
+                receiver.remaining_pairings(),
+            )
+            % receiver.remaining_pairings()
+        )
+        text += _("\nCancelling at this point will not use up a pairing.")
+    ok = prepare(receiver)
+    assistant = _create_assistant(receiver, ok, _finish, title, text)
+    if ok:
+        GLib.timeout_add(_STATUS_CHECK, _check_lock_state, assistant, receiver)
+    return assistant
 
-def _create_page(assistant, kind, header=None, icon_name=None, text=None):
-    p = Gtk.VBox(homogeneous=False, spacing=8)
-    assistant.append_page(p)
-    assistant.set_page_type(p, kind)
-
-    if header:
-        item = Gtk.HBox(homogeneous=False, spacing=16)
-        p.pack_start(item, False, True, 0)
-
-        label = Gtk.Label(label=header)
-        # deprecated - not needed label.set_alignment(0, 0)
-        label.set_line_wrap(True)
-        item.pack_start(label, True, True, 0)
-
-        if icon_name:
-            icon = Gtk.Image.new_from_icon_name(icon_name, Gtk.IconSize.DIALOG)
-            # deprecated - not needed icon.set_alignment(1, 0)
-            item.pack_start(icon, False, False, 0)
-
-    if text:
-        label = Gtk.Label(label=text)
-        # deprecated - not needed label.set_alignment(0, 0)
-        label.set_line_wrap(True)
-        p.pack_start(label, False, False, 0)
 
-    p.show_all()
-    return p
+def prepare(receiver):
+    if receiver.receiver_kind == "bolt":
+        if receiver.discover(timeout=_PAIRING_TIMEOUT):
+            return True
+        else:
+            receiver.pairing.error = "discovery did not start"
+            return False
+    elif receiver.set_lock(False, timeout=_PAIRING_TIMEOUT):
+        return True
+    else:
+        receiver.pairing.error = "the pairing lock did not open"
+        return False
 
 
 def _check_lock_state(assistant, receiver, count=2):
-    global address, kind, authentication, name, passcode
-
     if not assistant.is_drawable():
```

Comment:
Can this is_drawable call, be moved out of this function and done before, so check_lock_state call works without access to the UI?

Reply:
Done.